File: 2182.java
class Solution {
    public String repeatLimitedString(String s, int rl) {
        int f[] = new int[26];
        for (char c : s.toCharArray())
            ++f[c - 'a'];
        char res[] = new char[s.length()];
        int i = 0, j = 25, p = -1;
        while (j >= 0) {
            if (f[j] == 0)
                --j;
            else {
                if (p == j) {
                    int t = j - 1;
                    while (t >= 0 && f[t] == 0)
                        --t;
                    if (t == -1)
                        return new String(res, 0, i);
                    --f[t];
                    res[i++] = (char) ('a' + t);
                }
                int n = Math.min(f[j], rl);
                f[j] -= n;
                while (--n >= 0)
                    res[i++] = (char) ('a' + j);
                p = j;
            }
        }
        return new String(res, 0, i);
    }
}

File: 2182.py
class Solution:
    def repeatLimitedString(self, s: str, rl: int) -> str:
        f = [0] * 26
        for c in s:
            f[ord(c) - 97] += 1
        res = []
        i = 0
        j = 25
        p = -1
        while j >= 0:
            if f[j] == 0:
                j -= 1
            else:
                if p == j:
                    t = j - 1
                    while t >= 0 and f[t] == 0:
                        t -= 1
                    if t == -1:
                        return "".join(res)
                    f[t] -= 1
                    res.append(chr(t + 97))
                    i += 1
                n = min(f[j], rl)
                f[j] -= n
                while n > 0:
                    res.append(chr(j + 97))
                    n -= 1
                    i += 1
                p = j
        return "".join(res)
